core/blockchain.go

```diff
@@ -1198,7 +1212,19 @@ func (bc *BlockChain) WriteBlockWithState(
 	}
 
 	// Commit state object changes to in-memory trie
-	root, err := state.Commit(bc.chainConfig.IsS3(block.Epoch()))
+	var root common.Hash
+	skipValidators := bc.cacheConfig.Disabled && // archival node
+		!bc.chainConfig.IsAggregatedRewardEpoch(block.Epoch()) && // before aggregated reward epoch
+		block.NumberU64()%chain.RewardFrequency != chain.RewardFrequency-1 && // keep every 64th block
+		block.ShardID() == shard.BeaconChainShardID && // beacon chain
+		bc.chainConfig.IsPreStaking(block.Epoch()) // is actually staking
+	if skipValidators {
```

Comment:
similar comments as the other one. We don't want to modify the state calculation logic when doing the state commitment for the latest block, that's going to cause problem (missing validatorWrapper in the latest state). The last state should always have all the information and the logic for it shouldn't change. The only thing we want to change is to remove the history state data for validatorWrappers.

Reply:
As a example: https://medium.com/codechain/ethereums-state-trie-pruning-45ea73ed2c78#:~:text=In%20state%20trie%20pruning%2C%20it,the%20DB%20are%20not%20pruned.    we don't want to modify those blue blocks, we just want to remove the red blocks.